File: Solved Problems/sentimental-credit/credit.py
# TODO
from cs50 import get_string

def main():

    card = get_string("Give card nr\n")
    nrdig = len(card)
    if nrdig!= 16 and nrdig!= 15 and nrdig!= 13:
        print("INVALID")
    else:
        card = int(card)
        if checksum(card, nrdig)== 0:
            nrdig-=2
            card = card // 10**nrdig
            if amex(card)== 0:
                return 0
            if master(card)== 0:
                return 0
            card//= 10
            if visa(card,nrdig+2)== 0:
                return 0
            print("INVALID")
        else:
            print("INVALID")



def checksum(nr, digits):
    sum, odd, even = 0, 0, 0
    nr = int(nr)
    for i in range(digits):
        if i%2==0:
            odd+= nr%10
            nr//= 10
        else:
            temp= nr%10*2
            if temp>9:
                even+= temp%10
                temp//= 10
                even+= temp%10
            else:
                even+= temp%10
            nr//=10

    print(odd)
    sum= odd+even
    print(sum)
    if sum%10==0:
        return 0
    return 1

def amex(nr):

    if nr==34 or nr==37:
        print("AMEX")
        return 0
    return 1

def master(nr):

    if nr>=51 and nr<=55:
        print("MASTERCARD")
        return 0
    return 1

def visa(nr,digits):

    if nr==4:
        if digits== 13 or digits== 16:
            print("VISA")
            return 0
    return 1

if __name__ == "__main__":
    main()

File: Solved Problems/credit/credit.c
#include <cs50.h>
#include <stdio.h>

string checkamex(int nr)
{
    if (nr == 34)
    {
        return "AMEX";
    }
    if (nr == 37)
    {
        return "AMEX";
    }
    return "INVALID";
}
string checkmaster(int nr)
{
    if (nr <= 55 && nr >= 51)
    {
        return "MASTERCARD";
    }
    return "INVALID";
}
string checkvisa(int nr)
{
    if(nr/10 == 4)
    {
        return "VISA";
    }
    return "INVALID";
}
int prod(int n)
{
    n   *=2;
    int sum= n%10;
    n /= 10;
    sum += n%10;
    return sum;
}
int main(void)
{
    long nr = get_long("Give credit card number\n");
    string comp = "B";
    int sum = 0;
    int par = 1;
    int dig = 0;
    while (nr > 0)
    {
        if (par == 1)
        {
            sum += nr%10;
            nr /= 10;
            par = 2;
        }
        else
        {
            sum += prod(nr%10);
            nr/=10;
            par = 1;
        }
        if(nr<100 && comp[0] == 'B' )
        {
            comp = checkamex(nr);
            if(comp[0] == 'I')
                comp = checkmaster(nr);
            if(comp[0] == 'I')
                comp = checkvisa(nr);
        }
        dig ++;
    }
    if (sum%10 == 0 && dig>=13)
    {
        printf("%s\n",comp);
    }
    else {
        printf("INVALID\n");
    }
    return 0;

}